### PR TITLE
feat(expr): support interval `to_char()`

### PR DESCRIPTION
**Pull request by KeXiangWang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

To solve https://github.com/risingwavelabs/risingwave/issues/13561
Postgres' corresponding doc [here](https://www.postgresql.org/docs/current/functions-formatting.html).

In Postgres, an interval is first converted to a timestamp, and then use timestamp's `to_char()` implementation to convert the interval to a char. But `chrono` doesn't support interval type, especially negative intervals. 
 To reuse the format parsing implementation of `chrono`, in this PR, we still first convert the template to `chrono` template, and then format the interval based on the the `Item` types and the corresponding [behaviors](https://github.com/postgres/postgres/blob/8b965c549dc8753be8a38c4a1b9fabdb535a4338/src/backend/utils/adt/formatting.c#L2611C1-L2611C12) in Postgres. 

In the test file, you may see some output really wired. For example, 
```
select to_char('-120hours'::interval, 'HH12');
----
012
```
I have to say the implementation of postgres is not prudent and unified among different `Item`s. Here, I choose to strictly follow their behaviors.

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

### `to_char`

Converts the input to string according to the given format. Both uppercase and lowercase formats are supported.

```bash title=Syntax
to_char ( timestamptz, format ) → *string*
to_char ( timestamp, format ) → *string*
to_char ( interval, format ) → *string*
```

```bash title=Example

to_char(timestamp '2002-04-20 17:31:12.66', 'HH12:MI:SS') → '05:31:12'

to_char('2023-07-11 20:01:00-07:00'::timestamptz, 'HH12:MI:SS TZH:TZM') → 03:01:00 +00:00

to_char('1year 2 month 3day 4hour 5minute 6second'::interval, 'YYYY MM DD PM HH12 HH24 MI SS MS US') → 0001 02 03 AM 04 04 05 06 000 000000
```

### CODE REVIEW

**Review by xxchan — COMMENTED** [review]
Is there any pg regress tests for it? 🤔

**Comment by KeXiangWang** [comment]
> Is there any pg regress tests for it? 🤔

Good point! will check.

File: src/expr/impl/src/scalar/to_char.rs [hunk 1]
```diff
@@ -44,7 +54,7 @@ impl ChronoPattern {
     /// Compile the pg pattern to chrono pattern.
     // TODO: Chrono can not fully support the pg format, so consider using other implementations
     // later.
-    pub fn compile(tmpl: &str) -> ChronoPattern {
+    pub fn compile(tmpl: &str, is_interval: bool) -> ChronoPattern {
```
Comment (wangrunji0408) [inline src/expr/impl/src/scalar/to_char.rs:57]: Perhaps keep the original `compile` and add a new `compile_for_interval`?

File: src/expr/impl/src/scalar/to_char.rs [hunk 2]
```diff
@@ -138,3 +154,193 @@ fn timestamptz_to_char3(
...
+    "to_char(interval, varchar) -> varchar",
+    prebuild = "ChronoPattern::compile($1, true)"
+)]
+fn interval_to_char(
+    interval: Interval,
+    pattern: &ChronoPattern,
+    writer: &mut impl Write,
+) -> Result<()> {
+    for iter in pattern.borrow_dependent() {
+        println!("iter: {:?}", iter);
```
Comment (wangrunji0408) [inline src/expr/impl/src/scalar/to_char.rs:168]: debugging print should be removed

File: src/expr/impl/src/scalar/to_char.rs [hunk 3]
```diff
@@ -138,3 +154,193 @@ fn timestamptz_to_char3(
...
+            w.write_str(s).unwrap();
+            Ok(())
+        }
+        Item::Numeric(ref spec, _) => {
+            use chrono::format::Numeric::*;
+            match *spec {
+                Year => {
+                    let year = interval.years_field();
+                    if year < 0 {
+                        write!(w, "{:+01$}", year, 5).unwrap()
```
Comment (wangrunji0408) [inline src/expr/impl/src/scalar/to_char.rs:202]: Today I learned the "1$" format. But maybe embedding looks simpler? 
```suggestion
                        write!(w, "{:+05}", year).unwrap()
```

**Review by wangrunji0408 — COMMENTED** [review]
LGTM for the code. I don't check the format.

**Review by TennyZhuang — APPROVED** [review]
Generally LGTM

File: src/expr/impl/src/scalar/to_char.rs (hunk 1 above)
Comment (KeXiangWang) [reply src/expr/impl/src/scalar/to_char.rs]: done

File: src/expr/impl/src/scalar/to_char.rs (hunk 2 above)
Comment (KeXiangWang) [reply src/expr/impl/src/scalar/to_char.rs]: done

File: src/expr/impl/src/scalar/to_char.rs (hunk 3 above)
Comment (KeXiangWang) [reply src/expr/impl/src/scalar/to_char.rs]: Updated, Thx!

**Comment by KeXiangWang** [comment]
> Is there any pg regress tests for it? 🤔

Unfortunately, no available regress tests for us, the tests are either for `timestamp` and `timestamptz` or composed of currently unsupported patterns.

**Pull request merged**